### DISCUSSION TITLE
VSCode, Gradle init-script, custom repository

### DISCUSSION THREAD

**Discussion by moonkev** [scalameta/metals#2474] [Q&A]
Hello

When attempting to import a gradle project with metals in visual studio code, it seems metals generates a custom init-script with the bloop dependency setting the build script repository to maven central and attempts to run bloopInstall.  At my work we are behind a corporate proxy and do not have direct access to maven central repository from our machines.  We are forced to use artifactory to proxy in artifacts.  I already have the proper repository and bloop dependency setup in our build.gradle.  Is there any way to keep metals from trying to force this by creating and setting a --init-script when I import a project?

**Comment by tgodzik** [comment]
Thanks for the question! You can add the Bloop plugin manually and run `gradle bloopInstall` from console, which should have the same result. More information can be seen here: https://scalameta.org/metals/docs/build-tools/gradle.html#manual-installation We added the additional init script in order not to modify the build itself, but it seems in this case it's not possible. 

We should improve the support for proxies and mirrors. Most likely it should be something that we include in the settings itself. It would be super useful if you could explain how gradle proxies are defined. Is it via simple java properties?

**Reply by moonkev** [reply]
Hi @tgodzik Thank you for getting back, sorry slow to reply.  In gradle yes, proxies are definied using system properties.  These can be passed on the command line when invoking gradle or in a gradle.properties file placed in either your project folder or you home folder (if you want to apply globally to all projects).

It is the typical JDK proxy properties http.proxyHost, etc.

For my particular situation I am not using proxies. We are using an internal artifactory which does the proxying for us, so to configure that I just add as another normal repository

```groovy
repositories {
    maven {
        url "https://artifactory.company.com/release"
    }
}
```

Unfortunately even if install bloop manually, any time the project definition changes, VS Code automatically runs the --init-script one and breaks my setup.

Luckily bloop configuration in VS Code does allow you to specify a gradle executable.  So as a workaround I created a modified gradlew file which squashes out the --init-script parameter so that it regenerates using the plugin setup I have defined in my project's build.gradle.

**Reply by tgodzik** [reply]
We should most likely add repositories to the init script, but not sure exactly how to make it work. We do have custom repositories setting, which we could most likely use - we would need to translate it from the coursier land. Would you mind opening an issue?

Otherwise, your current workaround is probably best.

**Comment by moonkev** [comment]
Definitely, I will try to get an issue opened this evening.

**Reply by moonkev** [reply]
@tgodzik would you rather a feature request or bug request?

**Reply by tgodzik** [reply]
I would say it's a bug since it's preventing you from having it work reliably.
